File: relay/channel/codex/context_compaction.go
package codex

import (
	"fmt"
	"strings"

	"github.com/QuantumNous/new-api/common"
	"github.com/QuantumNous/new-api/dto"
)

const (
	defaultCompactionEnabled       = true
	defaultCompactionMinInputItems = 12
	defaultCompactionMaxInputItems = 24
	defaultCompactionKeepRecent    = 8
	defaultCompactionMaxSummary    = 4000
	defaultCompactionMaxItem       = 480
	compactionSummaryHeader        = "[Gateway compacted conversation summary]"
)

type contextCompactionSettings struct {
	Enabled             bool
	MinInputItems       int
	MaxInputItems       int
	PreserveRecentItems int
	MaxSummaryChars     int
	MaxItemChars        int
}

func clampInt(value, min, max int) int {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

func parseCompactionBool(value any, defaultValue bool) bool {
	switch typed := value.(type) {
	case bool:
		return typed
	case string:
		normalized := strings.TrimSpace(strings.ToLower(typed))
		switch normalized {
		case "1", "true", "yes", "on", "enabled", "enable", "auto":
			return true
		case "0", "false", "no", "off", "disabled", "disable", "none":
			return false
		}
	}
	return defaultValue
}

func parseCompactionInt(value any, defaultValue, min, max int) int {
	parsed, err := common.Any2Type[int](value)
	if err == nil {
		return clampInt(parsed, min, max)
	}
	if floatValue, err := common.Any2Type[float64](value); err == nil {
		return clampInt(int(floatValue), min, max)
	}
	if text, ok := value.(string); ok {
		return clampInt(common.String2Int(text), min, max)
	}
	return clampInt(defaultValue, min, max)
}

func parseContextCompactionSettings(raw []byte) contextCompactionSettings {
	settings := contextCompactionSettings{
		Enabled:             defaultCompactionEnabled,
		MinInputItems:       defaultCompactionMinInputItems,
		MaxInputItems:       defaultCompactionMaxInputItems,
		PreserveRecentItems: defaultCompactionKeepRecent,
		MaxSummaryChars:     defaultCompactionMaxSummary,
		MaxItemChars:        defaultCompactionMaxItem,
	}
	if len(raw) == 0 {
		return settings
	}

	switch common.GetJsonType(raw) {
	case "boolean":
		var enabled bool
		if err := common.Unmarshal(raw, &enabled); err == nil {
			settings.Enabled = enabled
		}
		return settings
	case "string":
		var text string
		if err := common.Unmarshal(raw, &text); err == nil {
			settings.Enabled = parseCompactionBool(text, defaultCompactionEnabled)
		}
		return settings
	case "object":
	default:
		return settings
	}

	var payload map[string]any
	if err := common.Unmarshal(raw, &payload); err != nil {
		return settings
	}
	if nested, ok := payload["compaction"].(map[string]any); ok {
		for key, value := range nested {
			payload[key] = value
		}
	}
	settings.Enabled = parseCompactionBool(payload["enabled"], defaultCompactionEnabled)
	mode := strings.TrimSpace(strings.ToLower(common.Interface2String(payload["mode"])))
	if mode == "" {
		mode = strings.TrimSpace(strings.ToLower(common.Interface2String(payload["type"])))
	}
	if mode == "" {
		mode = strings.TrimSpace(strings.ToLower(common.Interface2String(payload["strategy"])))
	}
	if mode == "disabled" || mode == "off" || mode == "none" {
		settings.Enabled = false
	}
	settings.MinInputItems = parseCompactionInt(payload["min_input_items"], defaultCompactionMinInputItems, 4, 256)
	if value, exists := payload["min_messages"]; exists {
		settings.MinInputItems = parseCompactionInt(value, settings.MinInputItems, 4, 256)
	}
	if value, exists := payload["min_items"]; exists {
		settings.MinInputItems = parseCompactionInt(value, settings.MinInputItems, 4, 256)
	}

	settings.MaxInputItems = parseCompactionInt(payload["max_input_items"], defaultCompactionMaxInputItems, 6, 512)
	if value, exists := payload["max_messages"]; exists {
		settings.MaxInputItems = parseCompactionInt(value, settings.MaxInputItems, 6, 512)
	}
	if value, exists := payload["max_items"]; exists {
		settings.MaxInputItems = parseCompactionInt(value, settings.MaxInputItems, 6, 512)
	}

	settings.PreserveRecentItems = parseCompactionInt(payload["preserve_recent_items"], defaultCompactionKeepRecent, 2, 128)
	if value, exists := payload["keep_recent_items"]; exists {
		settings.PreserveRecentItems = parseCompactionInt(value, settings.PreserveRecentItems, 2, 128)
	}
	if value, exists := payload["keep_recent"]; exists {
		settings.PreserveRecentItems = parseCompactionInt(value, settings.PreserveRecentItems, 2, 128)
	}

	settings.MaxSummaryChars = parseCompactionInt(payload["max_summary_chars"], defaultCompactionMaxSummary, 600, 24000)
	if value, exists := payload["summary_max_chars"]; exists {
		settings.MaxSummaryChars = parseCompactionInt(value, settings.MaxSummaryChars, 600, 24000)
	}

	settings.MaxItemChars = parseCompactionInt(payload["max_item_chars"], defaultCompactionMaxItem, 120, 2000)
	if value, exists := payload["item_max_chars"]; exists {
		settings.MaxItemChars = parseCompactionInt(value, settings.MaxItemChars, 120, 2000)
	}
	return settings
}

func safeJSON(value any) string {
	data, err := common.Marshal(value)
	if err != nil {
		return fmt.Sprintf("%v", value)
	}
	return string(data)
}

func truncateText(text string, limit int) string {
	if len(text) <= limit {
		return text
	}
	head := clampInt(limit/2, 32, limit)
	tail := clampInt(limit-head-16, 24, limit)
	if head+tail >= len(text) {
		return text
	}
	return text[:head] + " ...[trimmed]... " + text[len(text)-tail:]
}

func contentPartToText(part map[string]any) string {
	partType := strings.TrimSpace(strings.ToLower(common.Interface2String(part["type"])))
	switch partType {
	case "input_text", "output_text", "text", "summary_text":
		return common.Interface2String(part["text"])
	case "input_image":
		imageURL := common.Interface2String(part["image_url"])
		if strings.HasPrefix(imageURL, "data:") {
			return "[image:data-url]"
		}
		if imageURL != "" {
			return fmt.Sprintf("[image:%s]", imageURL)
		}
		return "[image]"
	case "input_file":
		if fileMap, ok := part["file"].(map[string]any); ok {
			fileName := common.Interface2String(fileMap["filename"])
			if fileName != "" {
				return fmt.Sprintf("[file:%s]", fileName)
			}
		}
		return "[file]"
	case "input_audio":
		return "[audio]"
	default:
		return ""
	}
}

func summarizeMessageItem(item map[string]any, maxItemChars int) string {
	role := strings.ToUpper(strings.TrimSpace(common.Interface2String(item["role"])))
	if role == "" {
		role = "USER"
	}
	content, _ := item["content"].([]any)
	textParts := make([]string, 0, len(content))
	for _, partAny := range content {
		part, ok := partAny.(map[string]any)
		if !ok {
			continue
		}
		text := contentPartToText(part)
		if text != "" {
			textParts = append(textParts, text)
		}
	}
	if len(textParts) == 0 {
		textParts = append(textParts, "[empty]")
	}
	return truncateText(role+": "+strings.Join(textParts, " "), maxItemChars)
}

func summarizeFunctionCall(item map[string]any, maxItemChars int) string {
	name := strings.TrimSpace(common.Interface2String(item["name"]))
	if name == "" {
		name = "tool"
	}
	callID := strings.TrimSpace(common.Interface2String(item["call_id"]))
	arguments := safeJSON(item["arguments"])
	prefix := "TOOL_CALL[" + name
	if callID != "" {
		prefix += "#" + callID
	}
	prefix += "]"
	return truncateText(prefix+": "+arguments, maxItemChars)
}

func summarizeFunctionResult(item map[string]any, maxItemChars int) string {
	callID := strings.TrimSpace(common.Interface2String(item["call_id"]))
	output := safeJSON(item["output"])
	prefix := "TOOL_RESULT"
	if callID != "" {
		prefix += "[" + callID + "]"
	}
	return truncateText(prefix+": "+output, maxItemChars)
}

func summarizeInputItem(item map[string]any, maxItemChars int) string {
	itemType := strings.TrimSpace(strings.ToLower(common.Interface2String(item["type"])))
	switch itemType {
	case "message":
		return summarizeMessageItem(item, maxItemChars)
	case "function_call":
		return summarizeFunctionCall(item, maxItemChars)
	case "function_call_output":
		return summarizeFunctionResult(item, maxItemChars)
	default:
		return truncateText(itemType+": "+safeJSON(item), maxItemChars)
	}
}

func fitSummaryLines(lines []string, maxChars int) string {
	if len(lines) == 0 {
		return ""
	}
	joined := strings.Join(lines, "\n")
	if len(joined) <= maxChars {
		return joined
	}
	headCount := clampInt(4, 1, len(lines))
	if headCount > len(lines) {
		headCount = len(lines)
	}
	head := append([]string(nil), lines[:headCount]...)
	marker := fmt.Sprintf("[...%d earlier items compacted...]", max(0, len(lines)-headCount))
	used := len(strings.Join(append(append([]string(nil), head...), marker), "\n"))
	tail := make([]string, 0, len(lines)-headCount)
	for idx := len(lines) - 1; idx >= headCount; idx-- {
		line := lines[idx]
		extra := len(line) + 1
		if used+extra > maxChars {
			break
		}
		tail = append([]string{line}, tail...)
		used += extra
	}
	candidate := append(head, marker)
	candidate = append(candidate, tail...)
	fitted := strings.Join(candidate, "\n")
	if len(fitted) <= maxChars {
		return fitted
	}
	return truncateText(fitted, maxChars)
}

func mergeInstructionsWithSummary(rawInstructions []byte, summaryText string, compactedCount, totalItems int) ([]byte, error) {
	summaryBlock := fmt.Sprintf(
		"%s\nCompacted earlier input items: %d of %d.\nTreat the summary below as background context. If it conflicts with the preserved recent turns, prefer the preserved recent turns.\n%s",
		compactionSummaryHeader,
		compactedCount,
		totalItems,
		summaryText,
	)
	var existing string
	if len(rawInstructions) > 0 {
		if err := common.Unmarshal(rawInstructions, &existing); err != nil {
			existing = strings.TrimSpace(string(rawInstructions))
		}
	}
	existing = strings.TrimSpace(existing)
	if existing != "" {
		return common.Marshal(existing + "\n\n" + summaryBlock)
	}
	return common.Marshal(summaryBlock)
}

func maybeCompactResponsesRequest(request *dto.OpenAIResponsesRequest) error {
	if request == nil {
		return nil
	}
	settings := parseContextCompactionSettings(request.ContextManagement)
	if !settings.Enabled {
		return nil
	}
	if common.GetJsonType(request.Input) != "array" {
		return nil
	}

	var inputItems []map[string]any
	if err := common.Unmarshal(request.Input, &inputItems); err != nil {
		return nil
	}
	if len(inputItems) == 0 {
		return nil
	}

	totalSerializedChars := 0
	for _, item := range inputItems {
		totalSerializedChars += len(safeJSON(item))
	}
	keepRecent := settings.PreserveRecentItems
	if keepRecent >= len(inputItems) {
		keepRecent = len(inputItems) - 1
	}
	if keepRecent < 1 {
		return nil
	}

	shouldCompact := len(inputItems) > settings.MaxInputItems ||
		(len(inputItems) >= settings.MinInputItems && totalSerializedChars > settings.MaxSummaryChars*2)
	if !shouldCompact {
		return nil
	}

	oldItems := inputItems[:len(inputItems)-keepRecent]
	recentItems := inputItems[len(inputItems)-keepRecent:]
	lines := make([]string, 0, len(oldItems))
	for _, item := range oldItems {
		line := summarizeInputItem(item, settings.MaxItemChars)
		if strings.TrimSpace(line) != "" {
			lines = append(lines, line)
		}
	}
	summaryText := fitSummaryLines(lines, settings.MaxSummaryChars)
	if strings.TrimSpace(summaryText) == "" {
		return nil
	}

	instructions, err := mergeInstructionsWithSummary(request.Instructions, summaryText, len(oldItems), len(inputItems))
	if err != nil {
		return err
	}
	encodedInput, err := common.Marshal(recentItems)
	if err != nil {
		return err
	}
	request.Input = encodedInput
	request.Instructions = instructions
	return nil
}

File: embedded-chatmock/chatmock/context_compaction.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Tuple


_DEFAULT_ENABLED = False
_DEFAULT_MIN_INPUT_ITEMS = 12
_DEFAULT_MAX_INPUT_ITEMS = 24
_DEFAULT_PRESERVE_RECENT_ITEMS = 8
_DEFAULT_MAX_SUMMARY_CHARS = 4000
_DEFAULT_MAX_ITEM_CHARS = 480
_SUMMARY_HEADER = "[Gateway compacted conversation summary]"


def _coerce_bool(value: Any, default: bool) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in ("1", "true", "yes", "on", "enabled", "enable", "auto"):
            return True
        if normalized in ("0", "false", "no", "off", "disabled", "disable", "none"):
            return False
    return default


def _coerce_int(value: Any, default: int, *, minimum: int, maximum: int) -> int:
    try:
        parsed = int(value)
    except Exception:
        parsed = default
    return max(minimum, min(maximum, parsed))


def _load_context_management(payload: Dict[str, Any] | None) -> Dict[str, Any]:
    raw = payload.get("context_management") if isinstance(payload, dict) else None
    if isinstance(raw, dict):
        nested = raw.get("compaction")
        if isinstance(nested, dict):
            merged = dict(raw)
            merged.update(nested)
            return merged
        return dict(raw)
    if isinstance(raw, bool):
        return {"enabled": raw}
    if isinstance(raw, str):
        return {"enabled": _coerce_bool(raw, _DEFAULT_ENABLED)}
    return {}


def _resolve_settings(payload: Dict[str, Any] | None) -> Dict[str, int | bool]:
    raw = _load_context_management(payload)
    enabled_default = _DEFAULT_ENABLED
    if isinstance(raw, dict) and raw:
        enabled_default = True
    enabled = _coerce_bool(raw.get("enabled"), enabled_default)
    mode = raw.get("mode") or raw.get("type") or raw.get("strategy")
    if isinstance(mode, str) and mode.strip().lower() in ("disabled", "off", "none"):
        enabled = False

    return {
        "enabled": enabled,
        "min_input_items": _coerce_int(
            raw.get("min_input_items") or raw.get("min_messages") or raw.get("min_items"),
            _DEFAULT_MIN_INPUT_ITEMS,
            minimum=4,
            maximum=256,
        ),
        "max_input_items": _coerce_int(
            raw.get("max_input_items") or raw.get("max_messages") or raw.get("max_items"),
            _DEFAULT_MAX_INPUT_ITEMS,
            minimum=6,
            maximum=512,
        ),
        "preserve_recent_items": _coerce_int(
            raw.get("preserve_recent_items") or raw.get("keep_recent_items") or raw.get("keep_recent"),
            _DEFAULT_PRESERVE_RECENT_ITEMS,
            minimum=2,
            maximum=128,
        ),
        "max_summary_chars": _coerce_int(
            raw.get("max_summary_chars") or raw.get("summary_max_chars"),
            _DEFAULT_MAX_SUMMARY_CHARS,
            minimum=600,
            maximum=24000,
        ),
        "max_item_chars": _coerce_int(
            raw.get("max_item_chars") or raw.get("item_max_chars"),
            _DEFAULT_MAX_ITEM_CHARS,
            minimum=120,
            maximum=2000,
        ),
    }


def _safe_dump(value: Any) -> str:
    try:
        return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    except Exception:
        return str(value)


def _truncate_text(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    head = max(32, limit // 2)
    tail = max(24, limit - head - 16)
    return text[:head] + " ...[trimmed]... " + text[-tail:]


def _content_part_to_text(part: Dict[str, Any]) -> str:
    part_type = str(part.get("type") or "").strip().lower()
    if part_type in ("input_text", "output_text", "text", "summary_text"):
        text = part.get("text")
        return text if isinstance(text, str) else ""
    if part_type == "input_image":
        image_url = part.get("image_url")
        if isinstance(image_url, str) and image_url.startswith("data:"):
            return "[image:data-url]"
        if isinstance(image_url, str) and image_url:
            return f"[image:{image_url}]"
        return "[image]"
    if part_type == "input_file":
        file_block = part.get("file")
        if isinstance(file_block, dict):
            file_name = file_block.get("filename")
            if isinstance(file_name, str) and file_name:
                return f"[file:{file_name}]"
        return "[file]"
    if part_type == "input_audio":
        return "[audio]"
    return ""


def _summarize_message_item(item: Dict[str, Any], max_item_chars: int) -> str:
    role = str(item.get("role") or "user").strip().upper()
    content = item.get("content")
    text_parts: List[str] = []
    if isinstance(content, list):
        for part in content:
            if not isinstance(part, dict):
                continue
            text = _content_part_to_text(part)
            if text:
                text_parts.append(text)
    elif isinstance(content, str) and content:
        text_parts.append(content)
    if not text_parts:
        text_parts.append("[empty]")
    return _truncate_text(f"{role}: {' '.join(text_parts)}", max_item_chars)


def _summarize_function_call(item: Dict[str, Any], max_item_chars: int) -> str:
    name = str(item.get("name") or "tool").strip()
    call_id = str(item.get("call_id") or item.get("id") or "").strip()
    arguments = item.get("arguments")
    arg_text = arguments if isinstance(arguments, str) else _safe_dump(arguments)
    prefix = f"TOOL_CALL[{name}"
    if call_id:
        prefix += f"#{call_id}"
    prefix += "]"
    return _truncate_text(f"{prefix}: {arg_text}", max_item_chars)


def _summarize_function_result(item: Dict[str, Any], max_item_chars: int) -> str:
    call_id = str(item.get("call_id") or item.get("id") or "").strip()
    output = item.get("output")
    output_text = output if isinstance(output, str) else _safe_dump(output)
    prefix = "TOOL_RESULT"
    if call_id:
        prefix += f"[{call_id}]"
    return _truncate_text(f"{prefix}: {output_text}", max_item_chars)


def _summarize_input_item(item: Dict[str, Any], max_item_chars: int) -> str:
    item_type = str(item.get("type") or "").strip().lower()
    if item_type == "message":
        return _summarize_message_item(item, max_item_chars)
    if item_type == "function_call":
        return _summarize_function_call(item, max_item_chars)
    if item_type == "function_call_output":
        return _summarize_function_result(item, max_item_chars)
    return _truncate_text(f"{item_type or 'item'}: {_safe_dump(item)}", max_item_chars)


def _fit_lines(lines: List[str], max_chars: int) -> str:
    if not lines:
        return ""
    joined = "\n".join(lines)
    if len(joined) <= max_chars:
        return joined

    head_count = min(4, len(lines))
    head = lines[:head_count]
    marker = f"[...{max(0, len(lines) - head_count)} earlier items compacted...]"
    used = len("\n".join(head + [marker]))
    tail: List[str] = []
    for line in reversed(lines[head_count:]):
        extra = len(line) + 1
        if used + extra > max_chars:
            break
        tail.insert(0, line)
        used += extra

    candidate = head + [marker] + tail
    fitted = "\n".join(candidate)
    if len(fitted) <= max_chars:
        return fitted
    return _truncate_text(fitted, max_chars)


def _build_summary_text(old_items: List[Dict[str, Any]], max_summary_chars: int, max_item_chars: int) -> str:
    lines = [
        _summarize_input_item(item, max_item_chars)
        for item in old_items
        if isinstance(item, dict)
    ]
    lines = [line for line in lines if line]
    return _fit_lines(lines, max_summary_chars)


def _merge_instructions(instructions: str | None, summary_text: str, compacted_count: int, total_items: int) -> str:
    summary_block = (
        f"{_SUMMARY_HEADER}\n"
        f"Compacted earlier input items: {compacted_count} of {total_items}.\n"
        "Treat the summary below as background context. If it conflicts with the preserved recent turns, "
        "prefer the preserved recent turns.\n"
        f"{summary_text}"
    )
    if isinstance(instructions, str) and instructions.strip():
        return instructions.rstrip() + "\n\n" + summary_block
    return summary_block


def maybe_compact_input_items(
    payload: Dict[str, Any] | None,
    input_items: List[Dict[str, Any]],
    instructions: str | None,
) -> Tuple[List[Dict[str, Any]], str | None, Dict[str, Any]]:
    settings = _resolve_settings(payload)
    meta: Dict[str, Any] = {
        "applied": False,
        "settings": settings,
        "original_items": len(input_items or []),
        "kept_items": len(input_items or []),
    }

    if not settings["enabled"]:
        meta["reason"] = "disabled"
        return list(input_items or []), instructions, meta

    if not isinstance(input_items, list) or not input_items:
        meta["reason"] = "empty"
        return list(input_items or []), instructions, meta

    total_serialized_chars = sum(len(_safe_dump(item)) for item in input_items)
    keep_recent = min(int(settings["preserve_recent_items"]), max(1, len(input_items) - 1))
    oversized_body = len(input_items) > 1 and total_serialized_chars > int(settings["max_summary_chars"]) * 2
    should_compact = len(input_items) > int(settings["max_input_items"]) or oversized_body or (
        len(input_items) >= int(settings["min_input_items"])
        and total_serialized_chars > int(settings["max_summary_chars"]) * 2
    )
    if not should_compact or keep_recent >= len(input_items):
        meta["reason"] = "below_threshold"
        return list(input_items), instructions, meta

    old_items = input_items[:-keep_recent]
    recent_items = input_items[-keep_recent:]
    summary_text = _build_summary_text(
        old_items,
        int(settings["max_summary_chars"]),
        int(settings["max_item_chars"]),
    )
    if not summary_text:
        meta["reason"] = "empty_summary"
        return list(input_items), instructions, meta

    meta.update(
        {
            "applied": True,
            "reason": "compacted",
            "compacted_items": len(old_items),
            "kept_items": len(recent_items),
            "summary_chars": len(summary_text),
        }
    )
    return recent_items, _merge_instructions(instructions, summary_text, len(old_items), len(input_items)), meta


def build_compaction_summary(
    payload: Dict[str, Any] | None,
    input_items: List[Dict[str, Any]],
) -> Tuple[str, Dict[str, Any]]:
    settings = _resolve_settings(payload)
    meta: Dict[str, Any] = {
        "settings": settings,
        "input_items": len(input_items or []),
    }
    if not isinstance(input_items, list) or not input_items:
        meta["reason"] = "empty"
        return "", meta

    summary_text = _build_summary_text(
        list(input_items),
        int(settings["max_summary_chars"]),
        int(settings["max_item_chars"]),
    )
    if not summary_text:
        meta["reason"] = "empty_summary"
        return "", meta
    meta["reason"] = "ok"
    meta["summary_chars"] = len(summary_text)
    return summary_text, meta
